flink-runtime/src/main/java/org/apache/flink/runtime/rpc/taskexecutor/ResourceManagerRegistration.java

```diff
@@ -0,0 +1,252 @@
+		this.resourceManagerLeaderId = checkNotNull(resourceManagerLeaderId);
+	}
+
+	// ------------------------------------------------------------------------
+	//  cancellation
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Cancels the registration procedure.
+	 */
+	public void cancel() {
+		canceled = true;
+	}
+
+	/**
+	 * Checks if the registration was canceled.
+	 * @return True if the registration was canceled, false otherwise.
+	 */
+	public boolean isCanceled() {
+		return canceled;
+	}
+
+	// ------------------------------------------------------------------------
+	//  registration
+	// ------------------------------------------------------------------------
+
+	/**
+	 * This method resolved the ResourceManager address to a callable gateway and starts the
+	 * registration after that.
+	 */
+	@SuppressWarnings("unchecked")
+	public void resolveResourceManagerAndStartRegistration() {
+		final RpcService rpcService = taskExecutor.getRpcService();
+
+		// trigger resolution of the resource manager address to a callable gateway
+		Future<ResourceManagerGateway> resourceManagerFuture =
+				rpcService.connect(resourceManagerAddress, ResourceManagerGateway.class);
+
+		// upon success, start the registration attempts
+		resourceManagerFuture.onSuccess(new OnSuccess<ResourceManagerGateway>() {
+			@Override
+			public void onSuccess(ResourceManagerGateway result) {
+				log.info("Resolved ResourceManager address, beginning registration");
+				register(result, 1, INITIAL_REGISTRATION_TIMEOUT_MILLIS);
+			}
+		}, taskExecutor.getMainThreadExecutionContext());
+
+		// upon failure, retry, unless this is cancelled
+		resourceManagerFuture.onFailure(new OnFailure() {
+			@Override
+			public void onFailure(Throwable failure) {
+				if (!isCanceled()) {
+					log.warn("Could not resolve ResourceManager address {}, retrying...", resourceManagerAddress);
+					resolveResourceManagerAndStartRegistration();
+				}
+			}
+		}, rpcService.getRpcExecutionContext());
+	}
+
+	/**
+	 * This method performs a registration attempt and triggers either a success notification or a retry,
+	 * depending on the result.
+	 */
+	@SuppressWarnings("unchecked")
+	private void register(final ResourceManagerGateway resourceManager, final int attempt, final long timeoutMillis) {
+		// this needs to run in the TaskExecutor's main thread
+		taskExecutor.validateRunsInMainThread();
+
+		// eager check for canceling to avoid some unnecessary work
+		if (canceled) {
+			return;
+		}
+
+		log.info("Registration at ResourceManager attempt {} (timeout={}ms)", attempt, timeoutMillis);
+
+		FiniteDuration timeout = new FiniteDuration(timeoutMillis, TimeUnit.MILLISECONDS);
+
+		Future<TaskExecutorRegistrationResponse> registrationFuture = resourceManager.registerTaskExecutor(
+				resourceManagerLeaderId, taskExecutor.getAddress(), taskExecutor.getResourceID(),
+				taskExecutor.getCurrentSlotReport(), timeout);
```

Comment:
Do we have to get the current slot report for each registration attempt? Isn't this something which can be out of date anyway when it arrives at the RM? The heartbeat should consolidate the state in such a case, shouldn't it?